### PR TITLE
[rllib] Modularize Torch and TF policy graphs

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

PyTorch:
- Split out the loss into a separate nn.Module
- Pull out a TorchPolicyGraph that handles the plumbing between the policy model and loss modules

TF:
- Clean up A3C graph def
- Pull out loss into a separate class when possible
- Clean up DDPG, DQN
- Leaving PPO for later cleanup

https://github.com/ray-project/ray/issues/2269

### CODE REVIEW

**Comment by ericl** [comment]
cc @richardliaw

File: python/ray/rllib/utils/torch_policy_graph.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        """Custom PyTorch optimizer to use."""
+        return torch.optim.Adam(self._model.parameters())
+
+    def compute_actions(
+            self, obs_batch, state_batches=None, is_training=False):
+        if state_batches:
+            raise NotImplementedError("Torch RNN support")
+        with self.lock:
+            ob = torch.from_numpy(np.array(obs_batch)).float()
+            model_out = self._model(ob)
```
Comment (vlad17) [inline python/ray/rllib/utils/torch_policy_graph.py]: with torch.no_grad():

Comment (vlad17) [reply python/ray/rllib/utils/torch_policy_graph.py]: on all of this, presumably

File: python/ray/rllib/utils/torch_policy_graph.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            logits = model_out[0]  # assume the first output is the logits
+            actions = F.softmax(logits, dim=1).multinomial(1).squeeze(0)
+            return var_to_np(actions), [], self.extra_action_out(model_out)
```
Comment (vlad17) [inline python/ray/rllib/utils/torch_policy_graph.py]: var_to_np impl should have .cpu() inside

File: python/ray/rllib/utils/torch_policy_graph.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                the loss. For example, ["obs", "action", "advantages"].
+
+        Attributes:
+            observation_space (gym.Space): observation space of the policy.
+            action_space (gym.Space): action space of the policy.
+        """
+        self.observation_space = observation_space
+        self.action_space = action_space
+        self.lock = Lock()
+        self._model = model
```
Comment (vlad17) [inline python/ray/rllib/utils/torch_policy_graph.py]: this module doesn't do any accounting for where the model is (i.e., it could be on a gpu device or cpu device). This only matters if the model is on a gpu and the from_numpy() inputs below give cpu tensors (Which they do).

With the code as-is there should be a check here that verifies model device == device('cpu')

Comment (vlad17) [reply python/ray/rllib/utils/torch_policy_graph.py]: or just docs saying input to the model is given as a cpu tensor

File: python/ray/rllib/utils/torch_policy_graph.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                loss_in.append(torch.from_numpy(postprocessed_batch[key]))
+            loss_in = [
+                t.float() if isinstance(t, torch.DoubleTensor) else t
+                for t in loss_in]
+            loss_out = self._loss(*loss_in)
+            self._optimizer.zero_grad()
+            loss_out.backward()
+            # Note that return values are just references;
+            # calling zero_grad will modify the values
+            return [p.grad.data.numpy() for p in self._model.parameters()], {}
```
Comment (vlad17) [inline python/ray/rllib/utils/torch_policy_graph.py]: do an if p.grad check inside the comprehension, and also use var_to_np

File: python/ray/rllib/utils/torch_policy_graph.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            actions = F.softmax(logits, dim=1).multinomial(1).squeeze(0)
+            return var_to_np(actions), [], self.extra_action_out(model_out)
+
+    def compute_gradients(self, postprocessed_batch):
+        with self.lock:
+            loss_in = []
+            for key in self._loss_inputs:
+                loss_in.append(torch.from_numpy(postprocessed_batch[key]))
+            loss_in = [
+                t.float() if isinstance(t, torch.DoubleTensor) else t
```
Comment (vlad17) [inline python/ray/rllib/utils/torch_policy_graph.py]: looks brittle

File: python/ray/rllib/utils/torch_policy_graph.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            loss_out.backward()
+            # Note that return values are just references;
+            # calling zero_grad will modify the values
+            return [p.grad.data.numpy() for p in self._model.parameters()], {}
+
+    def apply_gradients(self, gradients):
+        with self.lock:
+            self._optimizer.zero_grad()
+            for g, p in zip(gradients, self._model.parameters()):
+                p.grad = torch.from_numpy(g)
```
Comment (vlad17) [inline python/ray/rllib/utils/torch_policy_graph.py]: corresponding if not p.grad continue here. then you also wanto to copy_ to the data I think since you're feeding flat grads

File: python/ray/rllib/utils/torch_policy_graph.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            loss_out = self._loss(*loss_in)
+            self._optimizer.zero_grad()
+            loss_out.backward()
+            # Note that return values are just references;
+            # calling zero_grad will modify the values
+            return [p.grad.data.numpy() for p in self._model.parameters()], {}
+
+    def apply_gradients(self, gradients):
+        with self.lock:
+            self._optimizer.zero_grad()
```
Comment (vlad17) [inline python/ray/rllib/utils/torch_policy_graph.py]: looks unnecessary

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 1 above)
Comment (vlad17) [reply python/ray/rllib/utils/torch_policy_graph.py]: this == the forward pass

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 3 above)
Comment (ericl) [reply python/ray/rllib/utils/torch_policy_graph.py]: I think we want to support both, and just migrate the data automatically to the right place like TensorFlow does.
So for model output, it sounds like .cpu().numpy() will do. Is there an equivalent for moving cpu tensors to the model device?

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 5 above)
Comment (ericl) [reply python/ray/rllib/utils/torch_policy_graph.py]: Do you have a better idea? Ideally I'd insert an implicit cast to the right tensor type here.

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 4 above)
Comment (ericl) [reply python/ray/rllib/utils/torch_policy_graph.py]: You mean p.grad may be None?

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/utils/torch_policy_graph.py]: Makes sense.

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 6 above)
Comment (ericl) [reply python/ray/rllib/utils/torch_policy_graph.py]: Do you mean on `g`?

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 4 above)
Comment (vlad17) [reply python/ray/rllib/utils/torch_policy_graph.py]: depends -- sometimes you might freeze a param in which case p.grad would be None. but that's probably too much of an edge case to care about, you could just expect ppl to remove those from `parameters()`

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 3 above)
Comment (vlad17) [reply python/ray/rllib/utils/torch_policy_graph.py]: well, you should probably make var_to_np == .cpu().detach().numpy() and then just use that since you always need those steps.

Inputs are tricky. The model doesn't have an associated device, can be spread out across several devices, or whatever. I can see two ways of doing this, neither too great:
1. Doc that your model should accept CPU tensors, then in the fwd pass users should move to whatever device they think is appropriate.
2. Assume the model is on a single device, which is the `.device` of the model's first parameter (then move all inputs `.to()` that)

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 5 above)
Comment (vlad17) [reply python/ray/rllib/utils/torch_policy_graph.py]: sounds like it should be on the user to make the right numpy type, but I can see this doubletensor thing being a common bug. what do you do for TF?

File: python/ray/rllib/utils/torch_policy_graph.py (hunk 6 above)
Comment (vlad17) [reply python/ray/rllib/utils/torch_policy_graph.py]: well, as mentioned above just assuming p.grad is not NOne is probably fine.

then just do p.grad.data.copy_(torch.from_numpy(g))

Comment (ericl) [reply python/ray/rllib/utils/torch_policy_graph.py]: Hm yeah maybe we can assume not None for now. Btw when I tried copy_() sometimes p.grad.data was None and that failed.

**Review by ericl — COMMENTED** [review]
@vlad17 comments addressed.

File: python/ray/rllib/dqn/dqn_policy_graph.py [hunk 8]
```diff
@@ -17,32 +17,63 @@
...
-            if dones[i + j]:
-                break  # episode end
-    # truncate ends of the trajectory
-    new_len = len(obs) - n_step + 1
-    for arr in [obs, actions, rewards, new_obs, dones]:
-        del arr[new_len:]
+
+class QLoss(object):
+    def __init__(
+            self, q_t_selected, q_tp1_best, importance_weights, rewards,
```
Comment (richardliaw) [inline python/ray/rllib/dqn/dqn_policy_graph.py]: docs for these losses

File: python/ray/rllib/ddpg/ddpg_policy_graph.py [hunk 9]
```diff
@@ -22,62 +22,88 @@
...
+        for hidden in hiddens:
+            q_out = layers.fully_connected(
+                q_out, num_outputs=hidden, activation_fn=tf.nn.relu)
+        self.value = layers.fully_connected(
+            q_out, num_outputs=1, activation_fn=None)
+
+
+class ActorCriticLoss(object):
+    def __init__(
+            self, q_t, q_tp1, q_tp0, importance_weights, rewards, done_mask,
```
Comment (richardliaw) [inline python/ray/rllib/ddpg/ddpg_policy_graph.py]: doc this?

File: python/ray/rllib/dqn/dqn_policy_graph.py [hunk 10]
```diff
@@ -17,32 +17,63 @@
 Q_TARGET_SCOPE = "target_q_func"
 
 
-def adjust_nstep(n_step, gamma, obs, actions, rewards, new_obs, dones):
-    """Rewrites the given trajectory fragments to encode n-step rewards.
+class QNetwork(object):
+    def __init__(self, model, num_actions, dueling=False, hiddens=[256]):
+        with tf.variable_scope("action_value"):
+            action_out = model.last_layer
+            for hidden in hiddens:
```
Comment (richardliaw) [inline python/ray/rllib/dqn/dqn_policy_graph.py]: at some point, model specification should go through the model catalog

**Review by richardliaw — APPROVED** [review]
